pvlib/iotools/pvgis.py

```diff
@@ -0,0 +1,175 @@
+        north-west
+    startyear : integer [default None]
+        first year to calculate TMY
+    endyear : integer [default None]
+        last year to calculate TMY, must be at least 10 years from first year
+    url : str [default :const:`~pvlib.iotools.pvgis.URL`]
+        base url of PVGIS API, append ``tmy`` to get TMY endpoint
+    timeout : int, default 30
+        time in seconds to wait for server response before timeout
+
+    Returns
+    -------
+    data : pandas.DataFrame
+        the weather data
+    months_selected : list
+        TMY year for each month, ``None`` for basic and EPW
+    inputs : dict
+        the inputs, ``None`` for basic and EPW
+    meta : list or dict
+        meta data, ``None`` for basic
+
+    Raises
+    ------
+    requests.HTTPError
+        if the request response status is ``HTTP/1.1 400 BAD REQUEST``, then
+        the error message in the response will be raised as an exception,
+        otherwise raise whatever ``HTTP/1.1`` error occurred
+
+    """
+    # use requests to format the query string by passing params dictionary
+    params = {'lat': lat, 'lon': lon, 'outputformat': outputformat}
+    # pvgis only likes 0 for False, and 1 for True, not strings, also the
+    # default for usehorizon is already 1 (ie: True), so only set if False
+    if not usehorizon:
+        params['usehorizon'] = 0
+    if userhorizon is not None:
+        params['userhorizon'] = ','.join(str(x) for x in userhorizon)
+    if startyear is not None:
+        params['startyear'] = startyear
+    if endyear is not None:
+        params['endyear'] = endyear
+    res = requests.get(url + 'tmy', params=params, timeout=timeout)
+    # PVGIS returns really well formatted error messages in JSON for HTTP/1.1
+    # 400 BAD REQUEST so try to return that if possible, otherwise raise the
+    # HTTP/1.1 error caught by requests
+    if not res.ok:
+        try:
+            err_msg = res.json()
+        except Exception:
+            res.raise_for_status()
+        else:
+            raise requests.HTTPError(err_msg['message'])
+    if outputformat == 'json':
+        src = res.json()
+        return _parse_pvgis_tmy_json(src)
+    elif outputformat == 'csv':
+        src = io.BytesIO(res.content)
+        try:
+            data = _parse_pvgis_tmy_csv(src)
+        finally:
+            src.close()
+    elif outputformat == 'basic':
+        src = io.BytesIO(res.content)
+        try:
+            data = _parse_pvgis_tmy_basic(src)
+        finally:
+            src.close()
+    elif outputformat == 'epw':
+        src = io.StringIO(res.content.decode('utf-8'))
+        try:
+            data, meta = parse_epw(src)
+            data = (data, None, None, meta)
+        finally:
+            src.close()
+    else:
+        # this line is never reached because if outputformat is not valid then
+        # the response is HTTP/1.1 400 BAD REQUEST which is handled earlier
+        pass
```

Comment:
IMO this may fall into [EAFP](https://docs.python.org/3/glossary.html#term-eafp) _ie_ let's deal with that situation when it happens. My main concern is that there's no way to test raising an exception here because this line can't be reached as long as the pvgis api is working correctly.

Reply:
initialized data to all `None` which handles the possibility that the API could fail to handle a bad output format